[inductor] add conv_transpose2d unary fusion for cpu in inference mode

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90267
* #90266
* __->__ #90265
* #90264


An FX transformation is added to fuse ConvTranspose2d with eltwise OPs in torchinductor for CPU in inference mode, following the implementation in https://github.com/pytorch/pytorch/pull/87063.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/90264 and will be treated as an extern kernel call in torchinductor.

The fusion of ConvTranspose2d with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

chunyuan-w committed:
[inductor] add conv_transpose2d unary fusion for cpu in inference mode

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented on torch/_inductor/ir.py:
```diff
@@ -3411,6 +3411,70 @@ def _prepare_convolution_fusion_create(
 
 
+def _prepare_convolution_transpose_fusion_create(
+    cls,
```
Could we combine this code with convolution?

chunyuan-w replied:
Updated the code to reuse the `_prepare_convolution_fusion_create` function for `ConvTranspose`.

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on test/inductor/test_torchinductor.py:
```diff
@@ -1792,6 +1792,52 @@ def forward(self, x, y):
+            [1, 3],
+            [1, 2],
+            [1, 4],
+            [0],
```
Cover non-zero padding as well?

chunyuan-w replied:
Changed padding from `[0]` to `[0, 1]` to covert the non-zero padding case.

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"



An FX transformation is added to fuse ConvTranspose2d with eltwise OPs in torchinductor for CPU in inference mode, following the implementation in https://github.com/pytorch/pytorch/pull/87063.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/90264 and will be treated as an extern kernel call in torchinductor.

The fusion of ConvTranspose2d with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"



An FX transformation is added to fuse ConvTranspose2d with eltwise OPs in torchinductor for CPU in inference mode, following the implementation in https://github.com/pytorch/pytorch/pull/87063.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/90264 and will be treated as an extern kernel call in torchinductor.

The fusion of ConvTranspose2d with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"



An FX transformation is added to fuse ConvTranspose2d with eltwise OPs in torchinductor for CPU in inference mode, following the implementation in https://github.com/pytorch/pytorch/pull/87063.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/90264 and will be treated as an extern kernel call in torchinductor.

The fusion of ConvTranspose2d with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"



An FX transformation is added to fuse ConvTranspose2d with eltwise OPs in torchinductor for CPU in inference mode, following the implementation in https://github.com/pytorch/pytorch/pull/87063.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/90264 and will be treated as an extern kernel call in torchinductor.

The fusion of ConvTranspose2d with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"



An FX transformation is added to fuse ConvTranspose2d with eltwise OPs in torchinductor for CPU in inference mode, following the implementation in https://github.com/pytorch/pytorch/pull/87063.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/90264 and will be treated as an extern kernel call in torchinductor.

The fusion of ConvTranspose2d with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"



An FX transformation is added to fuse ConvTranspose2d with eltwise OPs in torchinductor for CPU in inference mode, following the implementation in https://github.com/pytorch/pytorch/pull/87063.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/90264 and will be treated as an extern kernel call in torchinductor.

The fusion of ConvTranspose2d with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"



An FX transformation is added to fuse ConvTranspose2d with eltwise OPs in torchinductor for CPU in inference mode, following the implementation in https://github.com/pytorch/pytorch/pull/87063.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/90264 and will be treated as an extern kernel call in torchinductor.

The fusion of ConvTranspose2d with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented:
@pytorchbot revert -c ghfirst -m "earlier pr on stack got yanked, this one needs to go too"

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"



An FX transformation is added to fuse ConvTranspose2d with eltwise OPs in torchinductor for CPU in inference mode, following the implementation in https://github.com/pytorch/pytorch/pull/87063.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/90264 and will be treated as an extern kernel call in torchinductor.

The fusion of ConvTranspose2d with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] add conv_transpose2d unary fusion for cpu in inference mode"



An FX transformation is added to fuse ConvTranspose2d with eltwise OPs in torchinductor for CPU in inference mode, following the implementation in https://github.com/pytorch/pytorch/pull/87063.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/90264 and will be treated as an extern kernel call in torchinductor.

The fusion of ConvTranspose2d with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w commented:
Re-opened in https://github.com/pytorch/pytorch/pull/91954
